src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,804 @@
+                        if (isAnyWorkerFutureLease)
+                        {
+                            string nextOwner = partition.NextOwner!;
+                            //If the NextOwner of the lease is the current worker, just plus 1 to the ownershipLeasesNum.
+                            if (nextOwner == this.workerName)
+                            {
+                                ownershipLeasesNum++;
+                                response.WaitForPartition = true;
+                            }
+                            //If the lease is stolen by other workers, add it to the partitionDistribution dictionary with NextOwner as key.
+                            else
+                            {
+                                if (partitionDistribution.ContainsKey(nextOwner))
+                                {
+                                    partitionDistribution[nextOwner].Add(partition);
+                                }
+                                else
+                                {
+                                    partitionDistribution.Add(nextOwner, new List<TableLease> { partition });
+                                }
+                            }
+                        }
+
+                        //If the lease belongs to a worker that is shutting down, steal it.
+                        if (isOtherWorkerShutDownLease)
+                        {
+                            oldOwner = partition.CurrentOwner!;
+                            this.StealLease(partition);
+                            isStealedLease = true;
+                            response.WaitForPartition = true;
+                        }
+                    }
+
+                }
+
+                // Balancing leases.
+                try
+                {
+                    await this.LeaseBalancer(partitionDistribution, partitions, ownershipLeasesNum, response);
+                }
+                catch (RequestFailedException ex) when (ex.Status == 412)
+                {
+                    throw;
+                }
+                
+
+                return response;
+
+            }
+
+            //This method is for the balance process.
+            //If is, then calculate the average number of leases per worker for balance.
+            //If ownership lease are less than the average number, then steal leases from other workers who have extra leases.
+            //The difference between workers's lease num should not exceed 1.
+            //Therefore, each worker should either have the average number of leases or one more than the average if the total number of leases cannot be evenly divided among them.
+            //Exception will be thrown if the update operation fails due to outdated etagso that worker could re-read the table again to get the latest information.
+            //Any other exceptions will be caught to logs. 
```

Comment:
Please consider alternative phrasing:

```suggestion
            // This method performs a worker-level partition rebalancing process.
            // The idea is to calculate the expected number of leases per worker in a fully balanced scenario
            //  (a.k.a their partition quota) and for each worker to steal partitions from others
            // until they have met their quota. 
            // A few remarks:
            // (1) The quota of partitions per worker is the number of partitions divided by the number of workers
            // (2) Workers only steal from workers that have exceeded their quota
```